File: examples/python/echo_server.py
#!/usr/bin/env python3
"""ROAR Protocol — Echo Server Example.

A minimal ROARServer that accepts DELEGATE messages and echoes them back.
Demonstrates Layer 1 (identity), Layer 4 (exchange), and Layer 2 (discovery registration).

Run this first, then run client.py in a second terminal.

Requirements:
    pip install -e ".[dev]"  (from the prowlrbot repo)
    pip install uvicorn fastapi  (for the HTTP server)

Usage:
    python3 examples/python/echo_server.py
"""

import asyncio
import logging

# ── Imports from ProwlrBot's ROAR reference implementation ──────────────────
from prowlrbot.protocols.roar import (
    AgentIdentity,
    MessageIntent,
    ROARMessage,
    StreamEvent,
    StreamEventType,
)
from prowlrbot.protocols.sdk.server import ROARServer

logging.basicConfig(level=logging.INFO, format="%(levelname)s  %(message)s")
log = logging.getLogger("echo-server")

# ── Step 1: Give this server an identity ────────────────────────────────────
identity = AgentIdentity(
    display_name="echo-server",
    agent_type="agent",
    capabilities=["echo", "reflect"],
)
log.info("Server DID: %s", identity.did)

# ── Step 2: Create the server ───────────────────────────────────────────────
server = ROARServer(
    identity=identity,
    host="127.0.0.1",
    port=8089,
    description="Echoes DELEGATE messages back to the sender",
    signing_secret="roar-example-shared-secret",
)


# ── Step 3: Register a handler for DELEGATE messages ────────────────────────
@server.on(MessageIntent.DELEGATE)
async def handle_delegate(msg: ROARMessage) -> ROARMessage:
    log.info(
        "← DELEGATE from %s: %s",
        msg.from_identity.display_name,
        msg.payload,
    )

    # Emit a StreamEvent so any subscriber can see this happened
    await server.emit(
        StreamEvent(
            type=StreamEventType.TASK_UPDATE,
            source=identity.did,
            session_id=msg.context.get("session_id", ""),
            data={"status": "echoed", "original_payload": msg.payload},
        )
    )

    response = ROARMessage(
        **{"from": server.identity, "to": msg.from_identity},
        intent=MessageIntent.RESPOND,
        payload={"echo": msg.payload, "status": "ok"},
        context={"in_reply_to": msg.id},
    )
    log.info("→ RESPOND: %s", response.payload)
    return response


# ── Step 4: Expose the handler over HTTP using FastAPI ───────────────────────
def create_app():
    """Create a FastAPI app that routes POST /roar/message to the server."""
    try:
        from fastapi import FastAPI, Request
        from fastapi.responses import JSONResponse
    except ImportError:
        raise ImportError("Install fastapi: pip install fastapi uvicorn")

    app = FastAPI(title="ROAR Echo Server")

    @app.post("/roar/message")
    async def receive_message(request: Request):
        body = await request.json()
        msg = ROARMessage(**body)

        # Verify HMAC signature (disable age check for demo)
        if msg.auth and not msg.verify("roar-example-shared-secret", max_age_seconds=0):
            return JSONResponse({"error": "invalid_signature"}, status_code=401)

        response = await server.handle_message(msg)
        return response.model_dump(by_alias=True)

    @app.get("/roar/agents")
    async def list_agents():
        card = server.get_card()
        return {"agents": [card.model_dump()]}

    return app


if __name__ == "__main__":
    try:
        import uvicorn
    except ImportError:
        print("Install uvicorn: pip install uvicorn")
        raise

    log.info("Starting ROAR echo server on http://127.0.0.1:8089")
    log.info("Endpoints:")
    log.info("  POST /roar/message  — receive a ROARMessage")
    log.info("  GET  /roar/agents   — see this server's AgentCard")
    log.info("")
    log.info("Now run: python3 examples/python/client.py")

    uvicorn.run(create_app(), host="127.0.0.1", port=8089, log_level="warning")

File: examples/ts/echo_server.ts
#!/usr/bin/env node
/**
 * ROAR Protocol — Echo Server (TypeScript/Node.js)
 *
 * Minimal HTTP server that accepts DELEGATE messages and echoes them back.
 * Demonstrates Layer 1 (identity), Layer 4 (exchange), and signing.
 *
 * Run this first, then run client.ts in a second terminal.
 *
 * Requirements:
 *   cd prowlrbot/packages/roar-sdk-ts && npm ci && npm run build
 *   # or use ts-node / Bun directly
 *
 * Usage (compiled):
 *   node dist/examples/ts/echo_server.js
 *
 * Usage (Bun):
 *   bun run examples/ts/echo_server.ts
 *
 * Usage (ts-node):
 *   npx ts-node examples/ts/echo_server.ts
 */

import * as http from "http";
import {
  MessageIntent,
  ROARMessage,
  createIdentity,
  createMessage,
  signMessage,
  verifyMessage,
} from "../../ts/src/index.js";

const PORT = 8089;
const SHARED_SECRET = "roar-example-shared-secret";

// ── Step 1: Give this server an identity ────────────────────────────────────
const identity = createIdentity("echo-server", {
  agentType: "agent",
  capabilities: ["echo", "reflect"],
});
console.log("Server DID:", identity.did);

// ── Step 2: Intent dispatch map ──────────────────────────────────────────────
async function handleMessage(incoming: ROARMessage): Promise<ROARMessage> {
  if (incoming.intent !== MessageIntent.DELEGATE) {
    return signMessage(
      createMessage(
        identity,
        incoming.from_identity,
        MessageIntent.RESPOND,
        {
          error: "unhandled_intent",
          message: `No handler for intent '${incoming.intent}'`,
        },
        { in_reply_to: incoming.id },
      ),
      SHARED_SECRET,
    );
  }

  console.log(
    `← DELEGATE from ${incoming.from_identity.display_name}:`,
    incoming.payload,
  );

  const response = signMessage(
    createMessage(
      identity,
      incoming.from_identity,
      MessageIntent.RESPOND,
      { echo: incoming.payload, status: "ok" },
      { in_reply_to: incoming.id },
    ),
    SHARED_SECRET,
  );

  console.log("→ RESPOND:", response.payload);
  return response;
}

// ── Step 3: Start the HTTP server ───────────────────────────────────────────
const server = http.createServer(async (req, res) => {
  if (req.method === "GET" && req.url === "/roar/health") {
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ status: "ok", protocol: "roar/1.0" }));
    return;
  }

  if (req.method !== "POST" || req.url !== "/roar/message") {
    res.writeHead(404, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ error: "not_found" }));
    return;
  }

  let body = "";
  req.on("data", (chunk) => (body += chunk));
  req.on("end", async () => {
    try {
      const data = JSON.parse(body) as ROARMessage;

      // Verify signature
      if (!verifyMessage(data, SHARED_SECRET)) {
        res.writeHead(403, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            error: "signature_invalid",
            detail: "HMAC signature verification failed.",
          }),
        );
        return;
      }

      const response = await handleMessage(data);
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(response));
    } catch (err) {
      res.writeHead(400, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({ error: "invalid_message", detail: String(err) }),
      );
    }
  });
});

server.listen(PORT, "127.0.0.1", () => {
  console.log(`\nROAR echo server listening on http://127.0.0.1:${PORT}`);
  console.log("Endpoints:");
  console.log("  POST /roar/message  — receive a ROARMessage");
  console.log("  GET  /roar/health   — health check");
  console.log("\nNow run: npx ts-node examples/ts/client.ts");
});
